src/Language/Haskell/LSP/Core.hs

```diff
@@ -316,70 +336,77 @@ mkSendReqFunc m params resHandler = do
     IsServerEither -> sendToClient $ FromServerMess m $ ReqMess msg
   return reqId
 
-newtype ClientRequestHandler (m :: Method FromClient Request)
-  = ClientRequestHandler
-  { runRequestHandler :: RequestMessage m -> HandlerFunc (ResponseParams m) -> IO ()
-  }
-
-newtype ClientNotificationHandler (m :: Method FromClient Notification)
-  = ClientNotificationHandler
-  { runNotificationHandler :: NotificationMessage m -> IO ()
-  }
-
-data Handlers
-  = Handlers
-  { requestHandlers :: DMap SMethod ClientRequestHandler
-  , notificationHandlers :: DMap SMethod ClientNotificationHandler
-  }
+type Handlers = forall t (m :: Method FromClient t). (SMethod m -> Maybe (Handler m))
 
 -- ---------------------------------------------------------------------
 nop :: Maybe (b -> LspM config ())
 nop = Nothing
 
 handlerMap :: (Show config) => SClientMethod m -> ClientMessage m -> LspM config ()
-handlerMap c = case c of
-  SWorkspaceDidChangeWorkspaceFolders -> hh (Just updateWorkspaceFolders) c
-  SWorkspaceDidChangeConfiguration    -> hh (Just handleConfigChange) c
-  STextDocumentDidOpen                -> hh (Just $ vfsFunc openVFS) c
-  STextDocumentDidChange              -> hh (Just $ vfsFunc changeFromClientVFS) c
-  STextDocumentDidClose               -> hh (Just $ vfsFunc closeVFS) c
-  SWorkDoneProgressCancel             -> hh (Just progressCancelHandler) c
-  _ -> hh nop c
+handlerMap m msg = do
+  -- First check to see if we have any dynamically registered handlers and call
+  -- their handlers
+  sf <- asks resSendMessage
+  dynReqHandlers <- readData resRegistrationsReq -- :: LspM config (RegistrationMap Request)
+  dynNotHandlers <- readData resRegistrationsNot
+  ~() <- case splitClientMethod m of
+    IsClientReq -> 
+      case DMap.lookup m dynReqHandlers of
+        -- TODO: If one request handles it, stop here and break out so we don't send back multiple responses?
+        Just (Compose regs) -> forM_ regs $ \(Pair _regId (RegistrationHandler h)) ->
+          liftIO $ h msg (mkClientResponseCallback sf msg)
```

Comment:
Do we really want to run `hh` to dispatch the static handlers after the message has been handled by dynamic handlers? I think we should only run `hh` if the dynamic handlers don't run.

Better yet, all of this logic should move to `hh`, as part of the code responsible for the handler lookup.